### PR TITLE
feat: Add API for ShardConfig

### PR DESCRIPTION
**Pull request by mkmik**
Part of #919

Mostly mechanical mapping between protobuf and internal rust types.

### CODE REVIEW

File: data_types/src/database_rules.rs [hunk 1]
```diff
@@ -800,6 +800,147 @@ impl PartialEq for Matcher {
...
+
+impl TryFrom<management::ShardConfig> for ShardConfig {
+    type Error = FieldViolation;
+
+    fn try_from(proto: management::ShardConfig) -> Result<Self, Self::Error> {
+        Ok(Self {
+            specific_targets: proto
+                .specific_targets
+                .map(|i| i.try_into())
+                .map_or(Ok(None), |r| r.map(Some))?,
```
Comment (alamb) [inline data_types/src/database_rules.rs:821]: To check my own understanding, is this the same as 

```suggestion
                .map(|i| i.try_into())?
```
 
I may be missing some subtlety

File: data_types/src/database_rules.rs [hunk 2]
```diff
@@ -800,6 +800,147 @@ impl PartialEq for Matcher {
...
+                .hash_ring
+                .map(|i| i.try_into())
+                .map_or(Ok(None), |r| r.map(Some))?,
+            ignore_errors: proto.ignore_errors,
+        })
+    }
+}
+
+/// Returns none if v matches its default value.
+fn default_option<T: Default + PartialEq>(v: T) -> Option<T> {
```
Comment (alamb) [inline data_types/src/database_rules.rs:832]: I wonder if this would more clearly be called `strip_default` or something?

**Review by alamb — APPROVED** [review]
Looks to me like it does what it says it does 👍

File: data_types/src/database_rules.rs (hunk 1 above)
Comment (carols10cents) [reply data_types/src/database_rules.rs:821]: No, the code you suggested there would try to return a `None` value from the whole function. It doesn't compile:

```
error[E0277]: `?` couldn't convert the error to `generated_types::google::FieldViolation`
   --> data_types/src/database_rules.rs:820:39
    |
820 |                 .map(|i| i.try_into())?,
    |                                       ^ the trait `From<std::option::NoneError>` is not implemented for `generated_types::google::FieldViolation`
```

(and isn't the behavior you want)

File: data_types/src/database_rules.rs [hunk 3]
```diff
@@ -800,6 +800,147 @@ impl PartialEq for Matcher {
...
+            node_groups: proto
+                .node_groups
+                .into_iter()
+                .map(try_from_management_node_group_for_node_group)
+                .collect::<Result<Vec<_>, _>>()?,
+        })
+    }
+}
+
+// cannot (and/or don't know how to) add impl From inside prost generated code
```
Comment (carols10cents) [inline data_types/src/database_rules.rs:894]: Yeah I think this is what has to happen

File: data_types/src/database_rules.rs [hunk 4]
```diff
@@ -800,6 +800,147 @@ impl PartialEq for Matcher {
...
+    Ok(proto.nodes.into_iter().map(|i| i.id).collect())
+}
+
+impl From<Matcher> for management::Matcher {
+    fn from(matcher: Matcher) -> Self {
+        Self {
+            table_name_regex: matcher
+                .table_name_regex
+                .map_or_else(|| "".into(), |r| r.to_string()),
+            predicate: matcher.predicate.unwrap_or_else(|| "".into()),
```
Comment (carols10cents) [inline data_types/src/database_rules.rs:916]: Not sure if this is worth it or not, but could these be:

```suggestion
            table_name_regex: matcher
                .table_name_regex
                .map(|r| r.to_string())
                .unwrap_or_default(),
            predicate: matcher.predicate.unwrap_or_default(),
```

I feel like it's a bit clearer what's going on? What do you think?

**Review by carols10cents — APPROVED** [review]
One tiny suggestion that I'm fine with either way!

File: data_types/src/database_rules.rs (hunk 2 above)
Comment (mkmik) [reply data_types/src/database_rules.rs]: opted for `none_if_default`

File: data_types/src/database_rules.rs (hunk 1 above)
Comment (mkmik) [reply data_types/src/database_rules.rs:821]: I learned this pattern (of turning an `Option<Result<...>>` into a `Result<Option<...`) from https://doc.rust-lang.org/beta/rust-by-example/error/multiple_error_types/option_result.html

File: data_types/src/database_rules.rs (hunk 4 above)
Comment (mkmik) [reply data_types/src/database_rules.rs:916]: thanks!

File: data_types/src/database_rules.rs (hunk 1 above)
Comment (alamb) [reply data_types/src/database_rules.rs:821]: I think you maybe can use https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose for a similar purpose

File: data_types/src/database_rules.rs (hunk 2 above)
Comment (alamb) [reply data_types/src/database_rules.rs]: that is better

File: data_types/src/database_rules.rs (hunk 1 above)
Comment (mkmik) [reply data_types/src/database_rules.rs:821]: Wow rust can have methods on specialization impls. I'm impressed.

**Pull request merged**